# 4. 회원 도메인 개발
## 03. 회원 기능 테스트
```java
@RunWith(SpringRunner.class)
@SpringBootTest
@Transactional
public class MemberServiceTest {

    @Autowired MemberService memberService;
    @Autowired MemberRepository memberRepository;

    @Test
    public void 회원가입() throws Exception {
        //given
        Member member = new Member();
        member.setName("kim");

        //when
        Long saveId = memberService.join(member);

        //then
        assertEquals(member, memberRepository.findOne(saveId));
    }

    @Test
    public void 중복_회원_예외() throws Exception {
        //given
        Member member1 = new Member();
        member1.setName("kim1");

        Member member2 = new Member();
        member2.setName("kim1");

        //when
        memberService.join(member1);
        try {
            memberService.join(member2);
        } catch (IllegalStateException e) {
            return;
        }

        //then
        fail("예외가 발생해야 한다.");
    }
}
```
### 스프링 통합 테스트
- @Runwith(SpringRunner.class): 스프링과 테스트 통합
- @SpringBootTest: 스프링 부트 띄우고 테스트(이게 없으면 @Autowired 다 실패)

### @Transactional
반복 가능한 테스트 지원, 각각의 테스트를 실행할 때마다 트랜잭션을 시작하고 테스트가 끝나면 트랜잭션을 강제로 롤백한다.

join() 메서드를 따라가보면 persist()만하지 commit()하는 부분이 없다. 
persist()의 INSERT 쿼리문은 commit()할 때 실행된다.
그러나, 트랜잭션이 롤백되면 영속성 컨텍스트가 flush하지 않는다.

롤백하지 않고 결과를 확인하고 싶다면 @Rollback(false) 애노테이션을 추가한다.
```java
@Test
@Rollback(false)
public void 회원가입() throws Exception { ... }
```
![image](https://github.com/GYUNGAEEEE/inflearn-SpringBoot-JPA/assets/158580466/e651d242-b6e4-4cc9-ad09-4f10b9202f2a)

DB에 저장된 것을 확인할 수 있다.

또는, 롤백을 하더라도 쿼리문을 실행한 것을 확인하고 싶다면 EntityManager를 주입받아 직접 flush 해준다.
```java
@Autowired EntityManager em;

@Test
public void 회원가입() throws Exception {

    ...
    //then
    em.flush();
    assertEquals(member, memberRepository.findOne(saveId));
}
```
```
insert 
into
    member
    (city, street, zipcode, name, member_id) 
values
    (?, ?, ?, ?, ?)
```
롤백되기 때문에 DB에 직접 저장되지는 않지만 INSERT 쿼리문을 확인할 수 있다.

### 중복 회원 예외
중복된 회원의 회원가입이 발생하면 IllegalStateException 예외가 발생한다.
중복 회원 예외 처리가 수행되지 않으면 fail() 메서드가 실행되고, 
fail() 메서드는 java.lang.AssertionError를 발생시키고 "예외가 발생해야 한다."라는 메시지를 출력한다.

try-catch 문으로 예외 처리를 하는 방법도 있고, @Test 애노테이션의 다음 옵션으로 예외 처리할 수 있다.
```java
@Test(expected = IllegalStateException.class)
```
